data/tutorials/language/1ms_01_functors.md

```diff
@@ -8,183 +8,381 @@ category: "Module System"
+  (public_name funkt)
+  (libraries str))
+```
+
+Check this works using the `dune exec funkt` command, it shouldn't do anything (the empty file is valid OCaml syntax) but it shouldn't fail either. The stanza `libraries str` will be used later.
+
+## Using an Existing Functor: `Set.Make`
+
+The standard library contains a [`Set`](/api/Set.html) module providing a data structure that allows operations like union and intersection. To use the provided type and its associated [functions](/api/Set.S.html), it is required to use the functor provided by `Set`. For reference only, here is a shortened version of the interface of `Set`.
+```ocaml
+module type OrderedType = sig
+  type t
+  val compare : t -> t -> int
+end
+
+module type S = sig
+  (* ... *)
+end
+
+module Make : functor (Ord : OrderedType) -> S
+```
+
+The functor `Set.Make` needs to be passed a module of type `Set.OrderedType` to produce a module of type `Set.S`. Here is how it can look like in our project:
+
+**`funkt.ml`**
+```ocaml
+module StringSet = Set.Make(String)
+```
+
+With this, the command `dune exec funkt` shouldn't do anything but it shouldn't fail either. Here is the meaning of that statement
+- The module `funkt.StringSet` is defined
+- The module `String` (from the standard library) is applied to the functor `Set.Make`. This is allowed because the `String` module satisfies the interface `Set.OrderedSet`
+  - It defines a type name `t` (which is an alias for `string`)
+  - It defines a function `compare` of type `t -> t -> bool`, that is the function `String.compare`
+- The result module from the functor application `Set.Make(String)` is bound to the name `StringSet`, it has the signature `Set.S`.
+
+Add some code to the `funkt.ml` file to produce an executable that does something and check the result.
+
+**`funkt.ml`**
+```ocaml
+module StringSet = Set.Make(String)
+
+let _ =
+  In_channel.input_lines stdin
+  |> List.concat_map Str.(split (regexp "[ \t.,;:()]+"))
+  |> StringSet.of_list
+  |> StringSet.iter print_endline
+```
+
+Here are the types of the functions used throughout the pipe
+- `In_channel.input_lines : in_channel -> string list`
+- `Str.(split (regexp "[ \t.,;:()]+")) : string -> string list`
+- `List.concat_map : ('a -> 'b list) -> 'a list -> 'b list`
+- `StringSet.of_list : string list -> StringSet.t` and
+- `StringSet.iter : StringSet.t -> unit`
+
+This reads the following way:
+- Read lines of text from standard input, produce a list of strings
+- Split each string using a regular expression, flatten the list of lists into a list
+- Convert the list of strings into a set
+- Display each element of the set
+
+The functions `StringSet.of_list` and `StringSet.iter` are available as the result of the functor application.
+
+```shell
+$ dune exec funkt < dune
+executable
+libraries
+name
+public_name
+str
+funkt
+```
+
+There are no duplicates in a `Set`. Therefore, the string `"funkt"` is only displayed once although it appears twice in the file `dune`.
```

Comment:
If we focus on the usage of the new set rather than input, regex splitting, etc, we wouldn't need much of this.

Reply:
You are right. I confess a preference for longer examples, but I see your point.